### PR TITLE
feat(templates): add a Fastify template

### PR DESCRIPTION
**Pull request by RinkiyaKeDad**
I've tried to remove express and apollo from the the mongodb template in an attempt to solve issue #2065. Due to my unfamiliarity with TypeScript and Fastify I'm facing trouble solving this problem:
ssTday

What I could understand from this is that I should not be passing ```mercurius``` in as an argument in app.register. But in first example of the [mercurius docs](https://github.com/mercurius-js/mercurius) it has been done in a similar way. Can you please point me in the right direction so that I can try to solve this error. Please also let me know if there are any other mistakes. Much thanks.

### CODE REVIEW

**Comment by wtrocki** [comment]
Mercurius is done and tested in JS. I would not be surprised to see some typescript issues. 
My recommendation would be to import this as nonts and simply make it work or use as any in place that is failing and point it out. Then I will see it myself and resolve that if possible or leave `as any`.

We can have separate PR for Mongo and Postgress migration ultimately. 
I think we need to rename the entire template and remove apollo from it as this will no longer use Apollo (it is using GraphQL-JS). Templates definitions are here: https://github.com/aerogear/graphback/blob/master/packages/create-graphback/src/init/starterTemplates.ts

If you willing to help us with this initiative and land numerous PR's for it let me know

**Comment by RinkiyaKeDad** [comment]
Thanks for responding @wtrocki. I'm all in for this! I'll start learning TS so that I'm able to contribute better 😅.

> My recommendation would be to import this as non ts and simply make it work or use as any in place that is failing and point it out. Then I will see it myself and resolve that if possible or leave `as any`.

Didn't get this. 


> to import this as non ts and simply make it work

I don't know how to do this. Can you point me towards where I can read more about it?



> or use as any in place that is failing and point it out

Didn't get this too 😅 I'm assuming it is a TS term?


> We can have separate PR for Mongo and Postgress migration ultimately.

Yup sounds good! Will work on the Postgres one after this.

> I think we need to rename the entire template

What do you mean by this? Didn't renaming the folder rename the template?



> remove apollo from it as this will no longer use Apollo

I'll remove it as a dependency in the ```ts-fastify-mongodb-backend``` folder.

**Comment by wtrocki** [comment]
> What do you mean by this? Didn't renaming the folder rename the template?

Rename folder with template and also template metadata. It will no longer apollo. 

As for typescript I think that repositories PR's are not the best place to teach it. 
Please take look on this: https://www.typescriptlang.org/docs/

**Comment by craicoverflow** [comment]
Hi @RinkiyaKeDad I got an error when I tried to run the template. Are you also experiencing this?

File: templates/ts-fastify-mongodb-backend/package.json [hunk 1]
```diff
@@ -14,15 +14,16 @@
   "license": "Apache 2.0",
   "dependencies": {
     "@graphback/runtime-mongo": "1.0.1",
-    "apollo-server-express": "2.17.0",
     "cors": "2.8.5",
     "dotenv": "8.2.0",
-    "express": "4.17.1",
+    "fastify": "^3.5.0",
```
Comment (wtrocki) [inline templates/ts-fastify-mongodb-backend/package.json:19]: Remove ^

File: templates/ts-fastify-mongodb-backend/src/index.ts [hunk 2]
```diff
@@ -35,16 +37,13 @@ async function start() {
...
   });
 
-  const apolloServer = new ApolloServer({
-    typeDefs,
-    resolvers: [resolvers, noteResolvers],
-    context: contextCreator
-  });
-
-  apolloServer.applyMiddleware({ app });
+  app.register((mercurius as any), {
```
Comment (wtrocki) [inline templates/ts-fastify-mongodb-backend/src/index.ts:40]: I think this kinda ended up being very complex PR for community member. 
We are missing subscription support. See: https://github.com/mercurius-js/mercurius

File: templates/ts-fastify-mongodb-backend/package.json (hunk 1 above)
Comment (RinkiyaKeDad) [reply templates/ts-fastify-mongodb-backend/package.json:19]: Didn't get you? :sweat_smile: Remove what? I've already removed express no?

**Review by machi1990 — CHANGES_REQUESTED** [review]
Hi @RinkiyaKeDad this is really good change. After some thoughts about this, here are few suggestions:

- Let's not move the existing mongo template to fastify as we are foreseeing some future changes in this template e.g bringing in Keycloak package etc, but also it acts as nice option for people looking for apollo template so let's keep it. Instead, we can re-purpose this PR to 
- Add a separate `ts-fastify-mongo-backend`  template. This is the entirety of your PR, it will be a matter of reusing some of this stuffs while keeping the old apollo-mongo template. This way, the new template can be tested easily and while at the same time bringing in the added option for people looking into a plain fastify-mongo-graphback template.   
- We can add a new entry to in [startertemplates.ts](https://github.com/aerogear/graphback/blob/ff151bbcae4ceb8bef3e637dc645e0a111534c20/packages/create-graphback/src/init/starterTemplates.ts) to include the ts-fastify-mongo-backend template. This can be based on: https://github.com/aerogear/graphback/blob/ff151bbcae4ceb8bef3e637dc645e0a111534c20/packages/create-graphback/src/init/starterTemplates.ts#L44-L52

Your suggestions are also welcomed. Thanks.

**Comment by RinkiyaKeDad** [comment]
Thanks for reviewing! I will add back the original template to this PR and make the required changes so that we'll be good to go.

**Comment by machi1990** [comment]
> Thanks for reviewing! I will add back the original template to this PR and make the required changes so that we'll be good to go.

Thank you a lot @RinkiyaKeDad

**Comment by RinkiyaKeDad** [comment]
I've added the mongodb template folder back and made the required changes to the ```graphback/packages/create-graphback/src/init/starterTemplates.ts``` file. Can you please review it now and suggest me what to do next?

File: templates/ts-fastify-mongodb-backend/README.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
+## Graphback MongoDB + Apollo Server Template
```
Comment (machi1990) [inline templates/ts-fastify-mongodb-backend/README.md:1]:
```suggestion
## Graphback MongoDB + Fastify Server Template
```

File: templates/ts-fastify-mongodb-backend/README.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+### Dependencies and Tools
+
+The following tools and technologies have been used to create this template:
+
+- [GraphQL](https://graphql.org/): GraphQL is an open-source data query and manipulation language for APIs which was publically released by Facebook in 2015.
+
+- [Apollo Server](https://www.apollographql.com/docs/apollo-server/): Apollo Server provides a way to simplify building the GraphQL server. It can be used with several popular libraries for Node.js like Express, Koa, Hapi. Here we have used it with Express.
+
+- [Express](https://expressjs.com/): Express is a minimal and flexible Node.js web application framework that makes building a Node.js server easier by providing a wide range of features.
```
Comment (machi1990) [inline templates/ts-fastify-mongodb-backend/README.md:40]: We are not using Apollo and Express, this section should talk about Fastify and Mecurus instead.

File: templates/ts-fastify-mongodb-backend/package.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "graphback-templates-mongodb-runtime",
+  "version": "1.0.1",
+  "description": "Graphback runtime template with Apollo Server and MongoDb",
```
Comment (machi1990) [inline templates/ts-fastify-mongodb-backend/package.json:4]: We are using Fastify in this template and not Apollo, let's change this to reflect that.

File: templates/ts-fastify-mongodb-backend/package.json [hunk 6]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "graphback-templates-mongodb-runtime",
```
Comment (machi1990) [inline templates/ts-fastify-mongodb-backend/package.json:2]:
```suggestion
  "name": "graphback-fastify-templates-mongodb-runtime",
```

File: templates/ts-fastify-mongodb-backend/src/resolvers/noteResolvers.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { QueryFilter } from "graphback";
+import { GraphQLResolveInfo } from 'graphql';
+import { IResolvers } from 'apollo-server-express';
```
Comment (machi1990) [inline templates/ts-fastify-mongodb-backend/src/resolvers/noteResolvers.ts:3]: We should get rid of this.

**Comment by RinkiyaKeDad** [comment]
I planned to update the readme once the code changes were done. @Machi should I now proceed to the readme now if no more changes need to be made to the code?

**Comment by machi1990** [comment]
> I've added the mongodb template folder back and made the required changes to the `graphback/packages/create-graphback/src/init/starterTemplates.ts` file. Can you please review it now and suggest me what to do next?

Thank you @RinkiyaKeDad I've added some comments, can you have a look?

**Comment by RinkiyaKeDad** [comment]
Much thanks for reviewing @machi1990! I've gone over your suggestions and implemented them. Please have a look.

**Comment by machi1990** [comment]
> Much thanks for reviewing @machi1990! I've gone over your suggestions and implemented them. Please have a look.

Thanks @RinkiyaKeDad for driving much of this work. However, the server could not be started and I realized a couple of missing things and I've created a commit to fix them - please take a look at https://github.com/aerogear/graphback/pull/2153/commits/00927d76b5bacaeb518849bc25f20bc71f2994d2. That should enable the server to startup.

@craicoverflow would you have sometime to do a check on this too? I'll be looking to merge this work today. And we can improve from there.

**Review by craicoverflow — APPROVED** [review]
Exemplary work and collaboration @RinkiyaKeDad!

**Comment by RinkiyaKeDad** [comment]
Thanks a lot for helping me out and giving me feedback :)

**Pull request merged**